src/hotspot/share/utilities/enumIterator.hpp

```diff
@@ -73,21 +74,59 @@
 // }
 
 // EnumeratorRange is a traits type supporting iteration over the enumerators of T.
-// Specializations must provide static const data members named
-// "_first" and "_last", whose values are the smallest / largest
-// (resp.) enumerator values for T. For iteration, the enumerators of
-// T must have sequential values in that range.
+// Specializations must provide static const data members named "_start" and "_end".
+// The type of _start and _end must be the underlying type of T.
+// _start is the inclusive lower bound of values in the range.
+// _end is the exclusive upper bound of values in the range.
+// The enumerators of T must have sequential values in that range.
 template<typename T> struct EnumeratorRange;
 
-// Specialize EnumeratorRange<T>.
-#define ENUMERATOR_RANGE(T, First, Last)        \
-  template<> struct EnumeratorRange<T> {        \
-    static constexpr T _first = First;          \
-    static constexpr T _last = Last;            \
+// Helper class for ENUMERATOR_RANGE and ENUMERATOR_VALUE_RANGE.
+struct EnumeratorRangeImpl : AllStatic {
+  template<typename T> using Underlying = std::underlying_type_t<T>;
+
+  template<typename T>
+  static constexpr Underlying<T> underlying(Underlying<T> value) {
+    return value;
+  }
+
+  // T not deduced to verify argument is of expected type.
+  template<typename T, typename U, ENABLE_IF(std::is_same<T, U>::value)>
+  static constexpr Underlying<T> start_value(U first) {
+    return static_cast<Underlying<T>>(first);
+  }
+
+  // T not deduced to verify argument is of expected type.
+  template<typename T, typename U, ENABLE_IF(std::is_same<T, U>::value)>
+  static constexpr Underlying<T> end_value(U last) {
+    Underlying<T> value = static_cast<Underlying<T>>(last);
+    assert(value < std::numeric_limits<Underlying<T>>::max(), "end value overflow");
+    return static_cast<Underlying<T>>(value + 1);
+  }
+};
+
+// Specialize EnumeratorRange<T>.  Start and End must be constant expressions
+// whose value is convertible to the underlying type of T.  They provide the
+// values of the required _start and _end members respectively.
+#define ENUMERATOR_VALUE_RANGE(T, Start, End)                           \
+  template<> struct EnumeratorRange<T> {                                \
+    static constexpr EnumeratorRangeImpl::Underlying<T> _start =        \
+      EnumeratorRangeImpl::underlying<T>(Start);                        \
+    static constexpr EnumeratorRangeImpl::Underlying<T> _end =          \
```

Comment:
Why do we need to call `underlying<T>(Start)` here? Can it be simplified as 
```
static constexpr EnumeratorRangeImpl::Underlying<T> _start = Start;
```
or
```
static constexpr EnumeratorRangeImpl::Underlying<T> _start {Start};
```

Reply:
I'd earlier considered requiring the type of Start and End be Underlying<T> rather than convertible to that.  The underlying() function is a partial remnant of that.  I'll remove it.

Reply:
Using brace initialization is the way to go, as it prevents narrowing.